src/cowboy_websocket.erl

```diff
@@ -411,3 +412,10 @@ handler_terminate(#state{env=Env, handler=Handler},
 		Req, HandlerState, Reason) ->
 	cowboy_handler:terminate(Reason, Req, HandlerState, Handler),
 	{ok, Req, [{result, closed}|Env]}.
+	
+	
+%% @private
+get_real_socket({proxy_socket, _, Socket, _, _, _, _, _, _}) -> Socket;
```

Comment:
I am still a bit too busy to get back to work just yet (and, to be frank, this is a topic that I will need time to get to in general, although I understand the problem with socket matching), but regardless this kind of patch will never be merged for the reasons @ferd mentioned.

My guts tell me extending the messages() function to return the socket part to match against is the right solution, *but* I don't think it's a good idea to take a decision on this until I get some practical experience with this (writing an example and experimenting things out) and until the changes to the HTTP/1.1 protocol get pushed, as they have more similar receive clauses and it won't be just Websocket anymore like this. And, similarly, the code to hook before Cowboy starts parsing HTTP will also be different. And then there's also the question of testing things out and such. It's such a large task that I didn't plan this for Cowboy 2 (CONNECT is in, but PROXY is still out) and while I don't rule out having some support in Cowboy 2, I believe we will need more time before PROXY support (or code to enable PROXY support and other) becomes stable.

Thoughts on messages(Socket) -> {OK, Closed, Error, MatchSocket}?